src/mito2/src/memtable/merge_tree/merger.rs

```diff
@@ -75,232 +66,173 @@ where
                             // with max sequence.
-                            let to_yield = top_node.slice_current_item(0..1);
-                            top_node.skip(1)?;
-                            to_yield
+                            self.current_rows = 1;
                         } else {
-                            let to_yield = top_node.slice_current_item(0..pos);
-                            top_node.skip(pos)?;
-                            to_yield
+                            // We don't know which one has the larger sequence so we use the range before
+                            // the duplicate pos.
+                            self.current_rows = pos;
                         }
                     }
                     Err(pos) => {
-                        // no duplicated timestamp
-                        let to_yield = top_node.slice_current_item(0..pos);
-                        top_node.skip(pos)?;
-                        to_yield
+                        // No duplication. Output rows before pos.
+                        debug_assert!(pos > 0);
+                        self.current_rows = pos;
                     }
-                };
-                self.current_item = Some(res);
+                }
             }
         } else {
-            // top is the only node left.
-            self.current_item = Some(top_node.fetch_next()?);
-        }
-        if top_node.is_valid() {
-            self.heap.push(top_node);
+            // Top is the only node left. We can read all rows in it.
+            self.current_rows = top_node.current_item_len();
         }
-        Ok(())
-    }
+        self.current_node = Some(top_node);
 
-    /// Returns current item held by merger.
-    pub(crate) fn current_item(&self) -> &T::Item {
-        self.current_item.as_ref().unwrap()
+        Ok(())
     }
-}
 
-#[derive(Debug)]
-pub struct DataBatchKey {
-    pk_index: PkIndex,
-    timestamp: i64,
-}
+    fn maybe_advance_current_node(&mut self) -> Result<()> {
+        let Some(mut node) = self.current_node.take() else {
+            return Ok(());
+        };
 
-impl Eq for DataBatchKey {}
+        // Advances current node.
+        node.advance(self.current_rows)?;
+        self.current_rows = 0;
+        if !node.is_valid() {
+            return Ok(());
+        }
 
-impl PartialEq<Self> for DataBatchKey {
-    fn eq(&self, other: &Self) -> bool {
-        self.pk_index == other.pk_index && self.timestamp == other.timestamp
-    }
-}
+        if self.heap.is_empty() {
+            // No other nodes in the heap, we can still put it into current node.
+            self.current_node = Some(node);
+        } else {
+            // Puts the node into the heap.
+            self.heap.push(node);
+        }
 
-impl PartialOrd<Self> for DataBatchKey {
-    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
-        Some(self.cmp(other))
+        Ok(())
```

Comment:
We can simplify this function to following, as the fast-path when heap is empty in which case `self.current_node` is set to current node happens only once per merger, but it brings more complexity.
```suggestion
        let Some(mut node) = self.current_node.take() else {
            return Ok(());
        };

        // Advances current node.
        node.advance(self.current_rows)?;
        self.current_rows = 0;
        if node.is_valid() {
            // Puts the node into the heap.
            self.heap.push(node);
        }
        Ok(())
```